pkg/forward/forward.go

```diff
@@ -478,53 +473,67 @@ func (isdf *InterStepDataForward) writeToBuffers(ctx context.Context, messageToS
 
 // writeToBuffer forwards an array of messages to a single buffer and is a blocking call or until shutdown has been initiated.
 func (isdf *InterStepDataForward) writeToBuffer(ctx context.Context, toBuffer isb.BufferWriter, messages []isb.Message) (writeOffsets []isb.Offset, err error) {
-	var totalBytes float64
-	for _, m := range messages {
-		totalBytes += float64(len(m.Payload))
-	}
-	writeOffsets = make([]isb.Offset, 0, len(messages))
-retry:
-	needRetry := false
+	var (
+		totalCount int
+		writeCount int
+		writeBytes float64
+		dropBytes  float64
+	)
+	totalCount = len(messages)
+	writeOffsets = make([]isb.Offset, 0, totalCount)
+
 	for {
 		_writeOffsets, errs := toBuffer.Write(ctx, messages)
 		// Note: this is an unwanted memory allocation during a happy path. We want only minimal allocation since using failedMessages is an unlikely path.
 		var failedMessages []isb.Message
-		for idx := range messages {
-			// use the messages index to index error, there is a 1:1 mapping
-			err := errs[idx]
-			// ATM there are no user defined errors during write, all are InternalErrors, and they require retry.
-			if err != nil {
-				needRetry = true
-				// we retry only failed messages
-				failedMessages = append(failedMessages, messages[idx])
-				writeMessagesError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": toBuffer.GetName()}).Inc()
-				// a shutdown can break the blocking loop caused due to InternalErr
-				if ok, _ := isdf.IsShuttingDown(); ok {
-					err := fmt.Errorf("writeToBuffer failed, Stop called while stuck on an internal error with failed messages:%d, %v", len(failedMessages), errs)
-					platformError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName}).Inc()
-					return writeOffsets, err
+		needRetry := false
+		for idx, msg := range messages {
+			if err := errs[idx]; err != nil {
+				// ATM there are no user defined errors during write, all are InternalErrors.
+				if _, ok := err.(isb.NoRetryableBufferWriteErr); ok {
```

Comment:
Use `errors.Is()`

Reply:
Nice, I didn't know about this. I changed it to use `errors.As()`, the difference is that `Is()` checks the error and target error as exactly same `==`, while `As()` checks that `the error's concrete value is assignable to the value pointed to by target error`, which is more close to what I want to do here - type check.